[19.05] Use byte comparison for contains and files_re_match

mvdbeek opened the pull request:
I don't think there's any drawback opening this in byte mode. ~~WIP because I want to add some tests before merging. Also the code looks more complicated than necessary.~~

Fixes https://github.com/galaxyproject/galaxy/issues/7957#issuecomment-493954901

mvdbeek committed:
Use byte comparison for contains

I don't think there's any drawback open this
in byte mode.

Fixes https://github.com/galaxyproject/galaxy/issues/7957#issuecomment-493954901

mvdbeek committed:
Unit test verify module

mvdbeek committed:
Use bytestring comparisons in verify module

mvdbeek committed:
Do non-shallow file comparison

nsoranzo commented on lib/galaxy/tools/verify/__init__.py:
```diff
@@ -279,37 +278,37 @@ def files_re_match(file1, file2, attributes=None):
     line_diff_count = 0
     while local_file:
-        contains = local_file.pop(0).rstrip('\n\r')
+        contains = local_file.pop(0).rstrip(b'\n\r')
```
This `while` looks weird, I'd replace the above 2 lines with:
```python
    for contains in local_file:
        contains = contains.rstrip(b'\n\r')
```

nsoranzo reviewed (commented):
I'm not sure I like this approach, in particular the bytes exception messages. Can we try to read the files as Unicode and if that fails, resort to binary mode?
So the exception messages can remain proper strings.

mvdbeek commented:
> I'm not sure I like this approach, in particular the bytes exception messages. Can we try to read the files as Unicode and if that fails, resort to binary mode ?

I get the hesitation, but reading data as unicode and then wrapping this with a byte fallback seems equally silly and makes for yet another path through the code that can fail.

> So the exception messages can remain proper strings.

If that's the concern we could also use unicodify on all Exceptions, do you think that'd be better ?

mvdbeek committed:
Use io.open with errors="repalce" instead of comparing byte strings

mvdbeek commented:
@nsoranzo I changed the file reading to `errors="replace"`, what do you think about that ?

nsoranzo commented:
> I changed the file reading to `errors="replace"`, what do you think about that ?

I guess it depends which problem are we trying to solve: if we want to allow proper checking of binary files (contain, regexps), then this won't work.

mvdbeek committed:
Work with bytestrings in `files_contains`, `files_re_match_*`

But read utf-8 unicode in `lines_diff`. This will replace non-utf8
characters and seems like a reasonable tradeoff for text files
mixed with byte contents, like pdfs.

mvdbeek commented:
I personally only want to fix https://github.com/galaxyproject/galaxy/issues/7957#issuecomment-493954901, but of course covering whatever makes sense. 

How about keeping contains and regex comparisons in bytes but do `lines_diff` in unicode ? I don't think `difflib.unified_diff` would work with non-utf8 characters (https://docs.python.org/3/library/difflib.html#difflib.unified_diff) which we could reasonably expect in some non-binary files (like pdf). Running `unicodify` would strip those characters and we'd still get a meaningful result.

I tried this in https://github.com/galaxyproject/galaxy/pull/8010/commits/a398127ede922d504ec7f1c39e66df22a9f0a16d

peterjc commented:
Unicode when looking at differences in lines makes sense (in this context should not have a binary file).

Raw bytes seems best when looking at file contains in a test (as used in the PDF example).

Also if we offered it, raw bytes for file starts with (that was the intent of my the PDF example - equally useful in PNG, SFF or something where there is a standard magic header).

nsoranzo committed:
Try to verify tool test output first as Unicode

nsoranzo commented:
@mvdbeek @peterjc I've pushed a commit with my idea of how to try to read the files first as Unicode and resort to binary mode if that fails with `UnicodeDecodeError`. I don't think it increases the code complexity too much, and it allows to do all the various checks for both text and binary files.

Let me know what do you think!

mvdbeek commented on lib/galaxy/tools/verify/__init__.py:
```diff
@@ -205,12 +206,21 @@ def get_lines_diff(diff):
+                history_data = [unicodify(l) for l in get_fileobj(file2, mode='rb', compressed_formats=compressed_formats)]
+                local_file = [unicodify(l) for l in get_fileobj(file1, mode='rb', compressed_formats=compressed_formats)]
+            else:
+                raise AssertionError("Binary data detected, not displaying diff")
```
This wouldn't work for the last example in the unit test if the extension isn't `.pdf`. This effectively makes non-utf8 encoded text uncomparable using `lines_diff`.  And it used to be comparable prior to our py3 porting.

mvdbeek replied:
Actually that's OK, we already failed the lines diff at this point ...

mvdbeek commented on lib/galaxy/tools/verify/__init__.py:
```diff
@@ -263,9 +273,21 @@ def get_lines_diff(diff):
-    history_data = open(file2, 'rb').readlines()
-    assert len(local_file) == len(history_data), 'Data File and Regular Expression File contain a different number of lines (%d != %d)\nHistory Data (first 40 lines):\n%s' % (len(local_file), len(history_data), ''.join(unicodify(history_data[:40])))
+    join_char = ''
+    to_strip = os.linesep
```
This isn't correct, we want to always strip carriage returns here, `\n` is already dealt with using `readlines`.

peterjc replied:
Do we care about Windows where ``os.linesep`` would be ``\r\n`` (CR LF)? Probably not.

However, more realistically, we probably should care about files using Windows new lines - the Python default universal text mode ought to take care of that case for us (although we'd have to double check Python 2 vs 3 there), that might be a good reason not to use binary mode for the line-based ``re`` check?

mvdbeek replied:
We actually want to always strip `\r\n` here, independent of what platform we're on. If old test data ends with `\r\n` and new data produced by a tool run is `\n` we don't want to fail IMO. We haven't in the past, and I don't think we should introduce this now.

mvdbeek replied:
Actually `\r\n` is being stripped by readlines, it'll just end up as `\n`, so that's actually the correct change. ...

peterjc commented:
Noted as a line diff, but I would worry about Unix (CR) vs DOS/Windows (CR LF) line endings if you want to handle line-based regular expressions in binary mode.

Using text mode and letting Python handle the cross-platform new line conventions could be simpler?

mvdbeek committed:
Add carriage return unit test

peterjc commented:
With the explicit inclusion of both Unix and DOS/Windows new lines in the tests, I'm happy. Thanks!

mvdbeek commented:
@nsoranzo thanks for the commit, I was still making some wrong assumptions there. It's OK for me now.

nsoranzo commented:
@mvdbeek Thanks for double-checking and adding the carriage return test! I'll merge when tests pass.

peterjc commented:
Will we need another pull request to get this work onto the ``dev`` branch (which is what I've been testing against with the Venn Diagram tool producing a PDF file)?

mvdbeek commented:
@nsoranzo merged 19.05 into dev in https://github.com/galaxyproject/galaxy/commit/0e14a47e6c29f78c8cfca478feb05c1f0d6d4ab9, so if you run the test it should hopefully pass now

peterjc commented:
Strange - I can see the changes now, but checked earlier and didn't. Most likely my mistake (although I saw a brief GitHub outage earlier today). Thanks.

mvdbeek commented:
Yeah, github's been showing some outdated changes yesterday as well.
I just ran the tests for https://github.com/peterjc/pico_galaxy/pull/33:
```
All 3 test(s) executed passed.
venn_list[0]: passed
venn_list[1]: passed
venn_list[2]: passed
```

So maybe that was just a intermittent issue ?

peterjc commented:
Fingers crossed - rerunning the TravisCI check now (it grabs Galaxy from the specified branch on GitHub).

peterjc commented:
Worked this time - thank you 👍
